### PR TITLE
March 30. Ruby agent: error fingerprinting and user tracking

### PR DESCRIPTION
**Pull request by fallwith**
The noticed errors documentation has been updated to explain how the error inbox related public API methods `set_error_group_callback` (for error fingerprinting) and `set_user_id` (for user tracking) work.

~**NOTE** This is a draft PR designed to accompany [Ruby agent #1833](https://github.com/newrelic/newrelic-ruby-agent/pull/1883). Once the Ruby agent PR has been merged and the Ruby agent team has slated the new code for an agent release, we'll have a better idea of when this docs PR needs to be reviewed and merged.~

### CODE REVIEW

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 1]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
     </tr>
   </tbody>
 </table>
+
+
+## Error fingerprinting: dynamically apply an error group to each noticed error
+
+A Proc based callback can be supplied to the agent to dynamically apply a desired [error group](../../../errors-inbox/errors-inbox) to each noticed error. Use the Ruby agent API [`NewRelic::Agent.set_error_group_callback`](https://www.rubydoc.info/github/newrelic/newrelic-ruby-agent/NewRelic/Agent#set_error_group_callback-instance_method) to provide the agent with a callback.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:106]:
```suggestion
A `Proc` based callback can be supplied to the agent to dynamically apply a desired [error group](../../../errors-inbox/errors-inbox) to each noticed error. Use the Ruby agent API [`NewRelic::Agent.set_error_group_callback`](https://www.rubydoc.info/github/newrelic/newrelic-ruby-agent/NewRelic/Agent#set_error_group_callback-instance_method) to provide the agent with a callback.
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 2]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
...
+
+This API call takes a callback function (must be of type `Proc`) as its only argument. The argument is required. The API call only needs to be made once per New Relic Ruby agent launch, so the call can be placed in a Rails initializer or whatnot. If subsequent calls to the API are made, they will update the callback to the newest one provided. Here is an example of a callback being defined and passed to the `NewRelic::Agent.set_error_group_callback` API method:
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:108]:
```suggestion
This API call takes a callback function (must be of type `Proc`) as its only argument. The argument is required. The API call only needs to be made once per New Relic Ruby agent launch, so the call can be placed in a Rails initializer or whatnot. If subsequent calls to the API are made, the callback function will update to the newest one provided. Here is an example of a callback being defined and passed to the `NewRelic::Agent.set_error_group_callback` API method:
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 3]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
...
+
+```
+proc = proc { |hash| "Access" if hash[:'http.statusCode'] == 401 }
+NewRelic::Agent.set_error_group_callback(proc)
+```
+
+In the example shown, a callback proc is created that will accept a hash as its only argument and then return the string "Access" for the desired error group name if the hash contains an HTTP status code key with a value of 401.
+
+The callback proc will be expected to receive exactly one input argument, a hash. The hash contains the following:
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:117]:
```suggestion
The callback proc is expected to receive exactly one input argument, a hash. The hash contains the following:
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 4]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
...
+      <td>`:'error.expected'`</td>
+      <td>Whether (true) or not (false) the error was expected</td>
+    </tr>
+    <tr>
+      <td>`:'options'`</td>
+      <td>The options hash passed to `NewRelic::Agent.notice_error`</td>
+    </tr>
+  </tbody>
+</table>
+```
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:161]:
```suggestion
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 5]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
...
+
+The callback proc will be expected to return a string representing the desired error group name if one can be determined. If the proc returns a `nil` or empty string (`''`), then the error will receive server-side grouping logic.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:163]:
```suggestion
The callback proc is expected to return a string representing the desired error group name if one can be determined. If the proc returns a `nil` or empty string (`''`), then the error will receive server-side grouping logic.
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx (hunk 2 above)
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:108]:
```suggestion
This API call takes a callback method (must be of type `Proc`) as its only argument. The argument is required. The API call only needs to be made once per New Relic Ruby agent launch, so the call can be placed in a Rails initializer or whatnot. If subsequent calls to the API are made, they will update the callback to the newest one provided. Here is an example of a callback being defined and passed to the `NewRelic::Agent.set_error_group_callback` API method:
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 6]
```diff
@@ -99,3 +99,74 @@ The `exception` is the exception to be recorded, or an error message. If needed,
...
+
+
+## User tracking: associating a user ID with each transaction and error
+
+Transactions and errors can be associated with a user ID if one is known to the New Relic Ruby agent. Use the Ruby agent API [`NewRelic::Agent.set_user_id`](https://www.rubydoc.info/github/newrelic/newrelic-ruby-agent/NewRelic/Agent#set_user_id-instance_method) to provide the agent with a user ID.
+
+This API call takes a single argument of a string representing a unique identifier for an end user. This string can be a UUID, a database id, etc. The argument is required. The API call should be made at least once per transaction to inform the New Relic Ruby agent of what user ID to associate the transaction with. Then in turn, when the agent notices errors during the lifespan of the transaction, the errors will bear an `enduser.id` agent attribute that holds the user ID value.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:170]:
```suggestion
This API call requires a single argument of a string representing a unique identifier for an end user. This string can be a UUID, a database id, etc. The API call should be made at least once per transaction to inform the New Relic Ruby agent of what user ID to associate the transaction with. Then in turn, when the agent notices errors during the lifespan of the transaction, the errors will bear an `enduser.id` agent attribute that holds the user ID value.
```

**Review by kaylareopelle — CHANGES_REQUESTED** [review]

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx [hunk 7]
```diff
@@ -99,3 +99,73 @@ The `exception` is the exception to be recorded, or an error message. If needed,
     </tr>
   </tbody>
 </table>
+
+
+## Error fingerprinting: dynamically apply an error group to each noticed error
+
+A `Proc` based callback can be supplied to the agent to dynamically apply a desired [error group](../../../errors-inbox/errors-inbox) to each noticed error. Use the Ruby agent API [`NewRelic::Agent.set_error_group_callback`](https://www.rubydoc.info/github/newrelic/newrelic-ruby-agent/NewRelic/Agent#set_error_group_callback-instance_method) to provide the agent with a callback.
+
+This API call takes a callback function (must be of type `Proc`) as its only argument. The argument is required. The API call only needs to be made once per New Relic Ruby agent launch, so the call can be placed in a Rails initializer or whatnot. If subsequent calls to the API are made, the callback function will update to the newest one provided. Here is an example of a callback being defined and passed to the `NewRelic::Agent.set_error_group_callback` API method:
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx:108]:
```suggestion
This API call takes a callback method (must be of type `Proc`) as its only argument. The argument is required. The API call only needs to be made once per New Relic Ruby agent launch, so the call can be placed in a Rails initializer or whatnot. If subsequent calls to the API are made, the callback method will update to the newest one provided. Here is an example of a callback being defined and passed to the `NewRelic::Agent.set_error_group_callback` API method:
```

File: src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx (hunk 2 above)
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/api-guides/sending-handled-errors-new-relic.mdx]: Superceded by [4cc5173](https://github.com/newrelic/docs-website/pull/12048/commits/4cc5173ffe7fefa1c728a2f814d4d6e8882e7058)

**Comment by fallwith** [comment]
NOTE: This PR should be merged and released in lockstep with https://github.com/newrelic/docs-website/pull/12357

**Review by zuluecho9 — APPROVED** [review]

**Pull request merged**